src/protocol/sort/mod.rs

```diff
@@ -137,3 +145,139 @@ impl AsRef<str> for MultiBitPermutationStep {
+    // 000 | 000 | 010 | 010 | 000 | 000 | 010 | 010
+    // 000 | 001 | 000 | 001 | 000 | 001 | 000 | 001
+    // 000 | 000 | 000 | 000 | 000 | 000 | 000 | 000
+    //
+    // Where 000, 101, 011, and 110 mean positive contributions, and
+    // 001, 010, 100, and 111 mean negative contributions.
+    let side_length = 1 << num_bits;
+    let mut equality_checks = Vec::with_capacity(side_length);
+    for i in 0..side_length {
+        let mut check = S::ZERO;
+        for (j, combination) in precomputed_combinations.iter().enumerate() {
+            let bit: i8 = i8::from((i & j) == i);
+            if bit > 0 {
+                if (!i & j).count_ones() & 1 == 1 {
+                    check -= combination;
+                } else {
+                    check += combination;
+                }
+            }
+        }
+        equality_checks.push(check);
+    }
+    Ok(equality_checks)
+}
+
+//
+// Every equality check can be computed as a linear combination of coefficients.
+// For example, if we are given a 3-bit number `[x_3, x_2, x_1]`,
+// we can check if it is equal to 4, by computing:
+// $x_3(1-x_2)(1-x_1)$,
+// which expands to:
+// $x_3 - x_2*x_3 - x_1*x_3 + x_1*x_2*x_3$
+//
+// Since we need to check all possible values, it makes sense to pre-compute all
+// of the coefficients that are used across all of these equality checks. In this way,
+// we can minimize the total number of multiplications needed.
+//
+// We must pre-compute all combinations of bit values. The following loop does so.
+// It does so by starting with the array `[1]`.
+// The next step is to multiply this by `x_1` and append it to the end of the array.
+// Now the array is `[1, x_1]`.
+// The next step is to mulitply all of these values by `x_2` and append them to the end of the array.
+// Now the array is `[1, x_1, x_2, x_1*x_2]`
+// The next step is to mulitply all of these values of `x_3` and append them to the end of the array.
+// Now the array is `[1, x_1, x_2, x_1*x_2, x_3, x_1*x_3, x_2*x_3, x_1*x_2*x_3]`
+// This process continues for as many steps as there are bits of input.
+async fn pregenerate_all_combinations<F, C, S>(
+    ctx: C,
+    record_idx: usize,
+    input: &[S],
+    num_bits: usize,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context,
+    S: SecretSharing<F> + BasicProtocols<C, F>,
+{
+    let record_id = RecordId::from(record_idx);
+    let mut precomputed_combinations = Vec::with_capacity(1 << num_bits);
+    precomputed_combinations.push(S::share_known_value(&ctx, F::ONE));
+    for (bit_idx, bit) in input.iter().enumerate() {
+        let step = 1 << bit_idx;
+        let mut multiplication_results =
+            try_join_all(precomputed_combinations.iter().skip(1).enumerate().map(
+                |(j, precomputed_combination)| {
+                    let child_idx = j + step;
+                    S::multiply(
+                        ctx.narrow(&BitOpStep::from(child_idx)),
```

Comment:
Doesn't this cause the BitOpStep to go out of range?

I think you need (2 ^ num_bits - 1) + 1 multiplications, and BitOpStep only has 64 strings. Also, this isn't exactly a bit-by-bit operation, so you might want to create a new step for this.

Reply:
It uses a lot fewer multiplications.

To check 4 requires 1 mult
To check 8 requires 4 mults
To check 16 requires 11 mults
To check 32 requires 26 mults
To check 64 requires 57 mults

It *will* break if you attempt to use this method for 65 or more breakdown keys, as that requires *120* multiplications per row...

...but the code is only invoked if you are using 32 or less breakdown keys.

I don't see the point in making yet another "Step" enum, as it won't really add any information.

Reply:
Ah, OK. I was worried that if the bit length is 7+ it'll break, but 32 was the cap.